File: samples/swift/Sources/PantherSDK.swift
import Foundation

public enum PantherSDK {

    public enum ProviderType: String, Codable {
        case openai
        case ollama
    }

    public struct LLM: Codable, Equatable {
        public var type: ProviderType
        public var baseURL: String
        public var model: String
        public var apiKey: String?

        public init(type: ProviderType, baseURL: String, model: String, apiKey: String? = nil) {
            self.type = type
            self.baseURL = baseURL
            self.model = model
            self.apiKey = apiKey
        }
    }

    public final class Client {
        private let providersJSON: String

        fileprivate init(providers: [LLM]) {
            _ = panther_init()
            let enc = JSONEncoder()
            enc.outputFormatting = [.withoutEscapingSlashes]
            let data = try? enc.encode(providers)
            self.providersJSON = String(data: data ?? Data("[]".utf8), encoding: .utf8) ?? "[]"
        }

        public func validate(prompt: String, guidelinesJSON: String? = nil) -> [String] {
            let raw = validateRawJSON(prompt: prompt, guidelinesJSON: guidelinesJSON)
            guard let data = raw.data(using: .utf8),
                  let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
                return [raw]
            }
            return arr.map { entry in
                let name = entry["provider_name"] as? String ?? "?"
                let score = (entry["adherence_score"] as? Double) ?? 0
                let latency = (entry["latency_ms"] as? Int) ?? 0
                return "\(name) – \(String(format: "%.1f", score))% – \(latency) ms"
            }
        }

        public func validateRawJSON(prompt: String, guidelinesJSON: String? = nil) -> String {
            if let guidelinesJSON = guidelinesJSON {
                guard let ptr = panther_validation_run_custom(prompt, providersJSON, guidelinesJSON) else {
                    return "{\"error\":\"ffi call failed\"}"
                }
                let json = String(cString: ptr)
                panther_free_string(ptr)
                return json
            } else {
                guard let ptr = panther_validation_run_multi(prompt, providersJSON) else {
                    return "{\"error\":\"ffi call failed\"}"
                }
                let json = String(cString: ptr)
                panther_free_string(ptr)
                return json
            }
        }
    }

    public static func make(llms: [LLM]) -> Client { Client(providers: llms) }

    // MARK: - Metrics & logs helpers
    public static func recordMetric(_ name: String, value: Double = 1.0) {
        _ = panther_metrics_record(name, value)
    }

    public static func listMetricNames() -> [String] {
        guard let ptr = panther_storage_list_metrics() else { return [] }
        let json = String(cString: ptr)
        panther_free_string(ptr)
        guard let data = json.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [String] else { return [] }
        return arr
    }

    public static func recentLogs() -> [String] {
        guard let ptr = panther_logs_get_recent() else { return [] }
        let json = String(cString: ptr)
        panther_free_string(ptr)
        guard let data = json.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [String] else { return [] }
        return arr
    }

    // MARK: - Guidelines cache & validation
    public enum Guidelines {
        private static let storageKey = "PantherSDK.GuidelinesJSON"
        public static let defaultJSON: String = """
\
[
  {
    \"topic\": \"Uso de medicamentos na gravidez\",
    \"expected_terms\": [
      \"consulta médica\",
      \"contraindicado\",
      \"categoria de risco\",
      \"dosagem\",
      \"efeitos colaterais\",
      \"advertência\",
      \"ANVISA\",
      \"orientação profissional\"
    ]
  }
]
"""
        public static func load() -> String {
            UserDefaults.standard.string(forKey: storageKey) ?? defaultJSON
        }
        public static func save(_ json: String) {
            UserDefaults.standard.set(json, forKey: storageKey)
        }
    }

    public static func validateGuidelineJSON(_ json: String) -> Bool {
        guard let data = json.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return false }
        return arr.allSatisfy { entry in
            guard let topic = entry["topic"] as? String,
                  let expected = entry["expected_terms"] as? [Any] else { return false }
            return !topic.isEmpty && expected.allSatisfy { $0 is String }
        }
    }
}

File: samples/kotlin/android/app/src/main/java/com/example/panther/PantherSDK.kt
package com.example.panther

import org.json.JSONArray
import org.json.JSONObject

class PantherSDK private constructor(private val providersJson: String) {

    data class LLM(
        val type: String,
        val base_url: String,
        val model: String,
        val api_key: String? = null,
    )

    fun validate(prompt: String, guidelinesJson: String? = null): List<String> {
        val raw = if (guidelinesJson != null) {
            PantherBridge.validateCustom(prompt, providersJson, guidelinesJson)
        } else {
            PantherBridge.validateMulti(prompt, providersJson)
        }
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                val name = o.optString("provider_name", "?")
                val score = o.optDouble("adherence_score", 0.0)
                val lat = o.optInt("latency_ms", 0)
                String.format("%s – %.1f%% – %d ms", name, score, lat)
            }
        } catch (_: Throwable) {
            listOf(raw)
        }
    }

    fun validateAndGetProof(prompt: String, guidelinesJson: String? = null): Pair<List<String>, String?> {
        val raw = if (guidelinesJson != null) {
            PantherBridge.validateCustomWithProof(prompt, providersJson, guidelinesJson)
        } else {
            PantherBridge.validateMultiWithProof(prompt, providersJson)
        }
        return try {
            val obj = org.json.JSONObject(raw)
            val arr = obj.getJSONArray("results")
            val tin = PantherBridge.tokenCount(prompt)
            val lines = (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                val name = o.optString("provider_name", "?")
                val score = o.optDouble("adherence_score", 0.0)
                val lat = o.optInt("latency_ms", 0)
                val text = o.optString("raw_text", "")
                val tout = PantherBridge.tokenCount(text)
                val rules = (customCostRulesJson ?: defaultCostRulesJson)
                val cost = PantherBridge.calculateCost(tin, tout, name, rules)
                String.format("%s – %.1f%% – %d ms – %d/%d tok – $%.4f", name, score, lat, tin, tout, cost)
            }
            val proof = obj.optJSONObject("proof")?.optString("combined_hash", null)
            lines to proof
        } catch (_: Throwable) {
            listOf(raw) to null
        }
    }

    companion object {
        val defaultGuidelines: String = """\
[
  {
    \"topic\": \"Uso de medicamentos na gravidez\",
    \"expected_terms\": [
      \"consulta médica\",
      \"contraindicado\",
      \"categoria de risco\",
      \"dosagem\",
      \"efeitos colaterais\",
      \"advertência\",
      \"ANVISA\",
      \"orientação profissional\"
    ]
  }
]
"""

        fun make(llms: List<LLM>): PantherSDK {
            PantherBridge.pantherInit()
            val arr = JSONArray()
            llms.forEach { l ->
                val o = JSONObject()
                o.put("type", l.type)
                o.put("base_url", l.base_url)
                o.put("model", l.model)
                if (l.type == "openai" && l.api_key != null) o.put("api_key", l.api_key)
                arr.put(o)
            }
            return PantherSDK(arr.toString())
        }

        fun version(): String = try { PantherBridge.version() } catch (_: Throwable) { "" }

        // Default pricing (editable in-app if desired)
        val defaultCostRulesJson: String = """
[
  {"match": "openai:gpt-4o-mini",  "usd_per_1k_in": 0.15, "usd_per_1k_out": 0.60},
  {"match": "openai:gpt-4.1-mini", "usd_per_1k_in": 0.30, "usd_per_1k_out": 1.20},
  {"match": "openai:gpt-4.1",      "usd_per_1k_in": 5.00,  "usd_per_1k_out": 15.00},
  {"match": "openai:gpt-4o",       "usd_per_1k_in": 5.00,  "usd_per_1k_out": 15.00},
  {"match": "openai:chatgpt-5",    "usd_per_1k_in": 5.00,  "usd_per_1k_out": 15.00},
  {"match": "ollama:llama3",       "usd_per_1k_in": 0.00,  "usd_per_1k_out": 0.00},
  {"match": "ollama:phi3",         "usd_per_1k_in": 0.00,  "usd_per_1k_out": 0.00},
  {"match": "ollama:mistral",      "usd_per_1k_in": 0.00,  "usd_per_1k_out": 0.00}
]
"""

        // Optional runtime override (set from activities/fragments)
        @JvmStatic var customCostRulesJson: String? = null
        @JvmStatic fun setCostRulesJson(json: String?) { customCostRulesJson = json }
    }
}
